### PR TITLE
Modularisation

### PR DESCRIPTION
**Pull request by mutlusun**
Hello,
this branch features the modularisation of the existing code. In the process I made some decisions that are debatable:

- I removed the openssl crate and used crates of the rust-crypto team. My intention was to reduce the number of crates that link to system libraries.
- I introducet `rusqlite` instead of `sqlite`. This fixes issue #10. I do not understand why the issue creator is so suspicious of this crate, but `rusqlite` is more often used, better maintained, and I think it was more easy to use in this context.
- I removed some of the command line arguments. This was mainly to ease development and I wasn't sure how useful they are. But they can be reintroduced easily.

However, I think all changes can now be implemented more easily. I'm not sure where to go from here. I would like to refactor the code further to implement some parallelisation. I don't think this crate is time critical but I think it could be a nice practice ...

Please add any suggestions!

Best

### CODE REVIEW

**Comment by mutlusun** [comment]
Hey @pajowu ,

Sorry for this large pull request. I assume it's hard to see which changes have happened. If you prefer to, I can split this large pull request into several smaller ones.

Best

**Comment by pajowu** [comment]
Sorry, i will take a close look in one week. I'm currently quite busy but should have more time then. Feel free to ping me if I didn't review it in 10 days

**Comment by pajowu** [comment]
I had the time to quickly skim through it and it looks very good so far. I saw some small things I want to have a look at before merging and will to a proper review next week. Thank you so much already :heart: 

One small request: could you rebase this branch to the current master / resolve the merge conflicts? If you don't find time now I can do it next week as well.

**Comment by mutlusun** [comment]
Thanks for your feedback! I rebased this branch and the branch can be merged now. No hurry and take your time ;)

**Comment by mutlusun** [comment]
Dear @pajowu ,
you want to take a look now? I have actually some async code already implemented in this branch as it does not change that much. Thank you ;) 
I think we should change the output though. Maybe we can use the [indicatif](https://crates.io/crates/indicatif) crate and output read and written bytes/frames.
Best

**Comment by mutlusun** [comment]
Hello,
I think the modularization branch is ready. I am in vacation starting from Saturday for two or three weeks. I won't be able to address any issues in this time.
Best

**Comment by pajowu** [comment]
I just took a look a this branch again. Looks good in general, but I have one thing that needs to be fixed before I will merge this: the construction the sqlite database in a temporary file (https://github.com/pajowu/signal-backup-decode/blob/master/src/main.rs#L479). I added this option, because the sqlite performance (at least on my machine) is horrible if the database file is not on a tmpfs. This seems to be because the data is immediately written to disk. I *think* we could work around this by using a `TRANSACTION` or similar, but I never got that working.

Could you change the code again so the database is constructed in a temp file again? If yes, please also add the `--no-tmp-sqlite` option, because moving the file does not work on some system configurations. If you don't want to do that, I can also add that

**Comment by mutlusun** [comment]
Thanks for your feedback!

I have wondered about the use of the `--no-tmp-sqlite` option and your explanation makes sense. I had the same problem that sqlite is really really slow but did not investigate it further. Instead of a temporary sqlite database I have added an in memory sqlite database. In the end, it is written to the actual output file.

The command line argument for it is added in the output formats branch.

File: src/output_raw.rs [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+			std::fs::create_dir_all(&path).with_context(|| {
+				format!("Path could not be created: {}", path.to_string_lossy())
+			})?;
+		}
+
+		// open database connection
+		let path_sqlite = path.join("signal_backup.db");
+
+		if path_sqlite.exists() && !force_write {
+			if force_write {
```
Comment (mutlusun) [inline src/output_raw.rs:46]: I have fixed this impossible condition in the `output-formats` branch.

**Comment by pajowu** [comment]
LGTM. When testing I noticed, that this branch is about 5-10 times slower while decrypting the database section of a backup and about 75 times slower when decrypting the attachment section. I'm gonna merge this now, but not craft a release. Feel free to profile it a bit to find out why it's so slow. I'm also gonna look into it if I find some time

**Comment by pajowu** [comment]
It seems to be the aes ctr library you added. I just tested it with the rust openssl bindings aes implementation and reached the old speed again.

**Pull request merged**